### ISSUE TITLE
Desynchronized onFocus and setState

### ISSUE THREAD

**Issue by murtraja** [retyui/react-native-confirmation-code-field#16]
This is an amazing library, however I have ran into a problem while making it work for web using react-native-web (it works perfectly fine on android)

The problem is that after inputting a character, it stays on the 1st Input. After inputting 2nd character, it goes to the next. I tried to understand the problem by keeping logs and i found the following

1. Input is given to the 1st widget
2. Forcing focus to be on the next (2nd) widget (<--- A)
3. onFocus again forces the focus to be on the 1st widget because state isn't updated (yet) (<---C)
4. The second step now continues and sets the state which was not updated above (<---B)

Please go through the logs below to better understand the situation.

```
[rnccf][handlerOnFocus] index:  0 
[rnccf][onKeyPress] 
[rnccf][onInputCode] character,index:  7 0 
[rnccf][onInputCode] calling setFocus                <--- A
[rnccf][setFocus] index:  1 
[rnccf][handlerOnFocus] index:  1 
[rnccf][handlerOnFocus] calling setFocus          <--- C
[rnccf][setFocus] index:  0 
[rnccf][handlerOnFocus] index:  0 
[rnccf][onInputCode] calling setState                 <--- B
[rnccf][onInputCode] setState completed
```

Clearly, the sequence should have been A then B then C.

Do you have any ideas regarding how can this be fixed for the web at the same time not breaking it for native? 

P.S. the complete code used to produce the logs:
<details><summary>Code</summary>
<p>

```
// @flow
import React, { PureComponent } from 'react';
import { View, TextInput as TextInputNative, StyleSheet } from 'react-native';
import PropTypes from 'prop-types';

import { getClassStyle, getInputSpaceStyle, isMatchingCode } from '../../utils';
import TextInput from '../TextInput';
import { getContainerStyle, styles } from './styles';
import { validateCompareCode, validateInputProps } from './validation';

import type { SyntheticEvent } from 'react-native/Libraries/Types/CoreEventTypes';
import type { VariantNames, INDEX } from '../../types';
import type { Props, State } from './types';

// eslint-disable-next-line
type DP = typeof ConfirmationCodeInput.defaultProps;

const getDefaultCodeSymbols: number => Array<string> = codeLength =>
  new Array(codeLength).fill('');

export default class ConfirmationCodeInput extends PureComponent<
  $Diff<Props, DP> & $Shape<DP>,
  State,
> {
  styles: Object;

  constructor(...args: any) {
    super(...args);

    const { defaultCode, codeLength, size } = this.props;

    this.state = {
      codeSymbols:
        typeof defaultCode === 'string'
          ? defaultCode.split('')
          : getDefaultCodeSymbols(codeLength),
      currentIndex: 0,
    };

    this.styles = StyleSheet.create({
      input: {
        width: size,
        height: size,
      },
      container: { height: size },
    });
  }

  onChangeCb = () => {
    //console.log("[rnccf][onChangeCb] ");
    const { onChangeCode } = this.props;

    if (onChangeCode) {
      onChangeCode(this.state.codeSymbols.join(''));
    }
  };

  clear() {
    this.setState(
      {
        currentIndex: 0,
        codeSymbols: getDefaultCodeSymbols(this.props.codeLength),
      },
      () => {
        this.setFocus(0);
      },
    );
  }

  detectFirstFocus: boolean = false;

  handlerOnFocus = (index: INDEX) => {
    console.log("[rnccf][handlerOnFocus] index: ", index);
    const newCodeArr = [...this.state.codeSymbols];
    const currentEmptyIndex = newCodeArr.findIndex(c => !c);

    if (currentEmptyIndex !== -1 && currentEmptyIndex < index) {
    console.log("[rnccf][handlerOnFocus] calling setFocus");
      return this.setFocus(currentEmptyIndex);
    }

    this.setState(
      {
        codeSymbols: newCodeArr.map((e, ind) => {
          if (ind >= index) {
            return '';
          }

          return e;
        }),
        currentIndex: index,
      },
      () => {
        if (this.detectFirstFocus) {
          this.onChangeCb();
        } else {
          this.detectFirstFocus = true;
        }
      },
    );
  };

  getClassStyle(variant: VariantNames, active: boolean) {
    const {
      activeColor,
      cellBorderWidth,
      inactiveColor,
      inputPosition,
      space,
    } = this.props;

    return getClassStyle(variant, active, {
      activeColor,
      cellBorderWidth,
      inactiveColor,
      inputSpaceStyle: getInputSpaceStyle(space, inputPosition),
    });
  }

  static IOS_MINIMAL_DELAY = 20;

  lastKeyEventTimestamp: number = 0;

  onKeyPress = (e: SyntheticEvent<*>) => {
    console.log("[rnccf][onKeyPress]");
    if (e.nativeEvent.key === 'Backspace') {
      /**
       * Due to a bug in RN iOS TextInput implementation, an unwanted backspace
       * event is fired on key press after clearing out the text input.
       * Typically this backspace event follows the actual event
       * and the time duration is under 10ms.
       * Added a check to see if we receive a backspace event under ~20ms
       * after the last key press event. If found, do nothing.
       *
       * Bug link: https://github.com/facebook/react-native/issues/18374
       */
      if (
        Math.abs(this.lastKeyEventTimestamp - e.timeStamp) <
        ConfirmationCodeInput.IOS_MINIMAL_DELAY
      ) {
        return;
      }

      const { currentIndex } = this.state;
      const nextIndex = currentIndex > 0 ? currentIndex - 1 : 0;

      this.setFocus(nextIndex);
    } else {
      // Record non-backspace key event time stamp
      this.lastKeyEventTimestamp = e.timeStamp;
    }
  };

  isLastIndex(index: INDEX): boolean {
    return index === this.props.codeLength - 1;
  }

  onInputCode = (character: string, index: INDEX) => {
    console.log("[rnccf][onInputCode] character,index: ", character, index);
    // on Android: text code is filled very slowly
    if (!this.isLastIndex(index)) {
      console.log("[rnccf][onInputCode] calling setFocus");
      this.setFocus(this.state.currentIndex + 1);
    }

    const {
      onFulfill,
      compareWithCode,
      ignoreCaseWhenCompareCode,
    } = this.props;
    const { currentIndex, codeSymbols } = this.state;
    const newCodeSymbols = [...codeSymbols];

    newCodeSymbols[index] = character;

    if (this.isLastIndex(index)) {
      const code = newCodeSymbols.join('');

      if (compareWithCode) {
        const isMatching = isMatchingCode(
          code,
          compareWithCode,
          ignoreCaseWhenCompareCode,
        );

        onFulfill(code, isMatching);

        if (isMatching) {
          this.clear();
        }
      } else {
        onFulfill(code);
      }

      this.blur(currentIndex);
    }

    console.log("[rnccf][onInputCode] calling setState");
    this.setState(prevState => ({
      codeSymbols: newCodeSymbols,
      currentIndex: prevState.currentIndex + 1,
    }), console.log("[rnccf][onInputCode] setState completed"))
  };

  codeInputRefs: Array<{ blur: () => void, focus: () => void }> = [];

  setFocus(index: INDEX) {
    console.log("[rnccf][setFocus] index: ", index);
    this.codeInputRefs[index].focus();
  }

  blur(index: INDEX) {
    this.codeInputRefs[index].blur();
  }

  setInputRef = (ref: any, idx: INDEX) => {
    this.codeInputRefs[idx] = ref;
  };

  getValue(value: string): string {
    const { maskSymbol } = this.props;

    return value ? maskSymbol || value.toString() : '';
  }

  renderInput(value: string, index: INDEX) {
    const {
      getInputStyle,
      autoFocus,
      variant,
      activeColor,
      getInputProps,
      keyboardType,
    } = this.props;
    const { currentIndex } = this.state;

    const customInputProps = getInputProps ? getInputProps(index) : null;

    if (process.env.NODE_ENV !== 'production') {
      validateInputProps(customInputProps);
    }

    const finalValue = this.getValue(value);

    return (
      <TextInput
        key={index}
        id={index}
        forwardRef={this.setInputRef}
        underlineColorAndroid="transparent"
        keyboardType={keyboardType}
        returnKeyType="done"
        selectionColor={activeColor}
        autoFocus={autoFocus && index === 0}
        value={finalValue}
        {...customInputProps}
        style={[
          styles.codeInput,
          this.styles.input,
          this.getClassStyle(variant, currentIndex === index),
          customInputProps && customInputProps.style,
          getInputStyle
            ? getInputStyle(index, currentIndex === index, Boolean(value))
            : null,
        ]}
        // fix for emoji  '😂'.length // 2
        maxLength={finalValue ? finalValue.length : 1}
        onChangeText={this.onInputCode}
        onFocus={this.handlerOnFocus}
        onKeyPress={this.onKeyPress}
      />
    );
  }

  render() {
    const { inputPosition, containerProps } = this.props;
    const { codeSymbols } = this.state;

    return (
      <View
        {...containerProps}
        style={[
          styles.container,
          getContainerStyle(inputPosition),
          this.styles.container,
          containerProps.style,
        ]}
      >
        {codeSymbols.map((value, id) => this.renderInput(value, id))}
      </View>
    );
  }

  static propTypes = {
    activeColor: PropTypes.string,
    autoFocus: PropTypes.bool,
    cellBorderWidth: PropTypes.number,
    codeLength: PropTypes.number,
    compareWithCode: validateCompareCode,
    containerProps: PropTypes.object,
    defaultCode: validateCompareCode,
    getInputProps: PropTypes.func,
    ignoreCaseWhenCompareCode: PropTypes.bool,
    inactiveColor: PropTypes.string,
    inputPosition: PropTypes.oneOf(['center', 'left', 'right', 'full-width']),
    onChangeCode: PropTypes.func,
    onFulfill: PropTypes.func.isRequired,
    size: PropTypes.number,
    space: PropTypes.number,
    variant: PropTypes.oneOf([
      'border-box',
      'border-circle',
      'border-b',
      'border-b-t',
      'border-l-r',
      'clear',
    ]),
    keyboardType: TextInputNative.propTypes.keyboardType,
    maskSymbol: PropTypes.string,
  };

  static defaultProps = {
    activeColor: '#fff',
    autoFocus: false,
    cellBorderWidth: 1,
    codeLength: 5,
    compareWithCode: null,
    containerProps: {},
    defaultCode: null,
    getInputProps: null,
    ignoreCaseWhenCompareCode: false,
    inactiveColor: '#ffffff40',
    inputPosition: 'center',
    onChangeCode: null,
    size: 40,
    space: 8,
    variant: 'border-box',
    keyboardType: 'default',
    maskSymbol: null,
  };
}
```
</p>
</details>

**Comment by retyui** [comment]
It's problem < RN 0.57 https://github.com/facebook/react-native/commit/5017b86

**Comment by murtraja** [comment]
I am using 
```
"react": "^16.6.0",
"react-art": "^16.6.0",
"react-dom": "^16.6.0",
"react-native": "0.57.3",
"react-native-web": "^0.9.6",
"react-native-confirmation-code-field": "^1.2.2",
```

I don't think it's because of slowness. I believe it is a design issue (synchronization). On digging further I found that why it works in Android but fails to work in web.
```
[rnccf][handlerOnFocus] index: ', 0
[rnccf][onInputCode] character,index: ', '7', 0
[rnccf][onInputCode] calling setFocus
[rnccf][setFocus] index: ', 1
[rnccf][onInputCode] calling setState
[rnccf][onInputCode] setState completed
[rnccf][onKeyPress]
[rnccf][handlerOnFocus] index: ', 1
```
In android it is always the case that setState call is completed and only after that onFocus is triggered (since we are using a JS Bridge, so all the work on JS gets finished before moving to another thread), but for the web, since there is no bridge, the onFocus can be called as soon as we call setFocus and hence onFocus uses the value of the state yet to be updated

**Comment by retyui** [comment]
Try to use `npm install react-native-confirmation-code-field@beta` new version

**Comment by murtraja** [comment]
I tried with the beta, facing same issue + the input is not displayed at all. Which means that I am never able to go to next widget. Atleast with the master, I was able to type one character, click on next widget, type another, click on next and so on. But here even that workaround is gone :)

**Comment by retyui** [comment]
Can you create a working demo with this problem?

**Comment by murtraja** [comment]
I spent some time in making this work. But all I could do was [this](https://codesandbox.io/s/14jr1kq7q7). The reason that it is so difficult to get a working demo lies in the config necessary to:

1. transpile the module to commonJS
2. alias react-native to react-native-web
3. alias react-native/Libraries to react-native-web/src

which can be easily done via a webpack config (the bundler for react apps)

Anyways, the other option is to fork the repo and do the necessary changes so that it works in an online environment. I will keep you posted if I am able to repro it. In the meantime, let me know if you find a solution or have doubts regarding my findings.

**Comment by murtraja** [comment]
Well as I mentioned earlier, I am making this work on web, so a snack wouldn't be helpful here for repro. Additionally, it works fine for native, so there isn't any point going there.

Also I have found a solution for master branch that works for native as well as web. In a nut shell, continuing the discussion from [above](https://github.com/retyui/react-native-confirmation-code-field/issues/16#issue-380557417), I simply delay calling setFocus (A) until the state is updated (C). So after this change, the flow is (input) -> (C) -> (A) -> (B)

I am not sure how would I go about this in the beta branch since I did not get enough time.
